### PR TITLE
Use grid for neighbour queries

### PR DESCRIPTION
**Pull request by Jononannes**

### CODE REVIEW

File: Assets/Scripts/Boid/BoidGrid.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            int zIndex = (int)(math.floor(info.pos.z) / cellDepth);
+            GridPoint gp = new GridPoint(xIndex, zIndex, cellXAmount);
+            //_grid.Add(gp, info);
+            _grid.Add(gp, i);
+        }
+    }
+
+
+    // Returns a complete map from boid index to list of neighbouring boid indices, for all boids
+    public NativeMultiHashMap<int, int> GetNeighbours()
```
Comment (palmdrop) [inline Assets/Scripts/Boid/BoidGrid.cs:35]: Is this method used other than in the test?

File: Assets/Scripts/Boid/Melee.cs [hunk 2]
```diff
@@ -26,14 +26,14 @@ public class Melee : Boid {
         collisionMask = LayerMask.GetMask("Wall", "Obstacle");
 
         classInfo = new ClassInfo {
-            viewRadius = 5f,
-            attackDstRange = 1f,
+            viewRadius = 1f,
+            attackDstRange = 0.5f,
```
Comment (palmdrop) [inline Assets/Scripts/Boid/Melee.cs:30]: Are the weights changed due to altered behaviors when using the grid, or was this for testing?

**Review by palmdrop — COMMENTED** [review]
Looks good. I was unsure about calculating all the boid neighbors all at once, but I suppose we'll need to do the same query for every boid, so it shouldn't matter. 

However, I'm thinking about how the grid is populated. Right now, this is done in the FlockStructJob. This job is not parallelized like the BoidStructJob is (which is a "IJobParallellFor" job). This way, the proximity queries are not run in parallel, if I understand everything correctly. Perhaps they could be? Instead of calculating one big MultiHashMap for all boid neighbors, perhaps it's better to just fetch the neighbors of each boid in parallel in the BoidStructJob?

File: Assets/Scripts/Boid/BoidGrid.cs (hunk 1 above)
Comment (Jononannes) [reply Assets/Scripts/Boid/BoidGrid.cs:35]: No, it is not. Should I remove it?

File: Assets/Scripts/Boid/Melee.cs (hunk 2 above)
Comment (Jononannes) [reply Assets/Scripts/Boid/Melee.cs:30]: The weights were changed for performance reasons. This is of course not optimal, but I think the separation behaviour has to be changed (increased) to fix this, or the cells will contain too many boids and there won't be a speedup.

**Comment by Jononannes** [comment]
@palmdrop I moved the usage of the grid to the parallelized job. This should help with performance, good catch.

File: Assets/Scripts/Boid/BoidGrid.cs (hunk 1 above)
Comment (palmdrop) [reply Assets/Scripts/Boid/BoidGrid.cs:35]: It's probably fine for now, but a bit unnecessary? If it's possible to run the test without this method, that might be better?

File: Assets/Scripts/Boid/Melee.cs (hunk 2 above)
Comment (palmdrop) [reply Assets/Scripts/Boid/Melee.cs:30]: Well no that's not ideal. There really should be a speedup anyway, unless the overhead of the grid is too large. Anything should be better than the naive O(n^2) approach

Comment (palmdrop) [reply Assets/Scripts/Boid/Melee.cs:30]: I think the view radius should be reduced actually. Ran some test in my own branch, and 2-3 might be a good radius. Right now, the groupings are way too large, which is one of the reasons they often just clump together.

**Review by palmdrop — COMMENTED** [review]
Looks good. Tested the code and seems to work alright. Would love to see how well it works with the improved behaviors before merging to main. Waiting for approval on that pull request.

Have you done any profiling? With the current cell size, is there a clear performance increase?

**Review by palmdrop — APPROVED** [review]
I resolved the merge conflicts and tried running the project. Seems to work as it should. Good work on the grid. I think we're ready to merge. Approved.

**Pull request merged**